util_function_groups/column_operations.R

```diff
@@ -0,0 +1,137 @@
+            # if it does, throw a warning if the column is required but it doesn't exist
+            if(column_attributes$column_required & !column_attributes$column %in% df_names){
+                util.warn("Required column " %+% column %+% " does not appear in the data.")
+            }
+        }
+        print("Validated required")
+    }
+    
+    for(column in names(column_validation)){
+        
+        # pick out the attributes of the column in question
+        column_attributes <- column_validation[[column]]
+        
+        # add the column name itself as an attribute of the column
+        column_attributes$column <- column
+        
+        # check whether the column exists
+        if(column %in% names(df)){
+            values <- df[,column]
+            # If it does, run the appropriate checks
+            for(attribute in names(column_attributes)){
+                switch(attribute,
+                       blanks_allowed = validate_blanks_allowed(column_attributes,values),
+                       accepted_values = validate_accepted_values(column_attributes,values),
+                       accepted_range = validate_accepted_range(column_attributes,values),
+                       datatype = validate_datatype(column_attributes,values)
+                )
+            }
+        }
+        validate_required(column_attributes, names(df))
+        # Finally, check that the column exists, if it's required
+        # If it was, throw a warning.
+    }
+}
+
+################################ Unit testing ####################################
+
+# @todo - turn into unit tests
+
+d <- read.csv("~/Documents/ctc_intervention.csv", stringsAsFactors = FALSE) %>%
+    qc.clean_qualtrics() %>% # clean the qualtrics
+    # add the test columns
+    mutate(
+        blanks_in_wrong_place = sample(c(1:7, NA), nrow(.), replace = TRUE),
+        out_of_range = sample(c(1:8, NA), nrow(.), replace = TRUE),
+        blanks_not_among_accepted_values = sample(c("Value1", "Value2", NA),
+                                                  nrow(.), replace = TRUE)
+    )
+
+column_validation <- list(
+    # some archetype columns
+    "nonexistant_required" = list(
+        "column_required" = TRUE
+    ),
+    "nonexistant_not_required" = list(
+        "column_required" = FALSE
+    ),
+    "blanks_in_wrong_place" = list(
+        "blanks_allowed" = FALSE
+    ),
+    "out_of_range" = list(
+        "accepted_range" = c(1, 7)
+    ),
+    "blanks_not_among_accepted_values" = list(
+        "accepted_values" = c("Value1", "Value2")
+    ),
+    # some real columns from Qualtrics:
+    "race" = list(
+        "datatype" = "character",
+        "accepted_values" = c("White", "Black", "Latino", "Asian", "Other", "Unknown"),
+        "column_required" = FALSE,
+        "blanks_allowed" = TRUE
+    ),
+    "ddowell" = list(
+        "datatype" = "numeric",
+        "accepted_range" = c(1, 7),
+        "blanks_allowed" = TRUE
```

Comment:
hmm, I didn't think about blanks in json, just R. And I used the util.is_blank() function to find the blanks, which would not pick out a `null` value:
function(x){
  is.na(x) | grepl("^[ \t]*$", x)
}
It seems like making this whole thing compatible with json output is a big source of value, so probably worth it to add `null` to the list of blank values. Honestly, could we do that by adding `null` to util.is_blank, i.e.,
  is.na(x) | grepl("^[ \t]*$", x) | x %in% "null"
I can't imagine a case where we'd have a value in a cell that's literally the string "null" and we wouldn't want to just call it blank.